### PR TITLE
Upgrade to Kodi 20

### PR DESCRIPTION
**Pull request by sanzoghenzo**
Hi there!
I almost finished the rewrite of the addon to support kodi 20.

I still have to figure out why the "firstRun" settings is not persisted between restarts (it always displays the discovery dialog, even if it's already configured).

It surely has other bugs around the code because my hyperion is inside a container and the uPnP discovery doesn't work, so I couldn't test it.

To describe what I did:
- I split all the various responsibilities into separate classes, and used Dependency Injection (DI) to lessen the coupling between the classes (this is also good for unit testing, but I still didn't implement any test :sweat_smile: );
- The Player and Monitor classes only notify the Hyperion class with a specific command (so that the hyperion class doesn't need to access either of them in order to get the kodi state);
- I added two methods to ApiClient (former Connection) to check if hyperion needs the api authentication and ask for a token, but they're not used and tested - I'll try to make them work on the next iteration
- Settings are handled by the SettingsManager class;
- logging is handled by the Logger class;
- gui module is now a class, it still needs some more DI love, but it's working;
-  I've added the Pipfile to be able to setup a development environment with pipenv;
- I've added a github workflow to check the addon at each push and send a PR to the official kodi repository on a new tag creation.

### CODE REVIEW

**Comment by sanzoghenzo** [comment]
I suspect there's a bug on the new `Settings` wrapper class, I opened an [issue on the kodi repo](https://github.com/xbmc/xbmc/issues/23551).

In the meantime, I'll work around this problem by using the old `addon.setSettingsXxxx`.

**Comment by Lord-Grey** [comment]
@sanzoghenzo Thank you very much for your efforts bringing the add-on to the state of the art.
Very much appreciated.

I provided you some additional feedback on Discord esp. in regard of discovery. 
Have you had a chance looking at it?
If you prefer Feedback only via the PR, I can provide here too.

**Comment by sanzoghenzo** [comment]
> @sanzoghenzo Thank you very much for your efforts bringing the add-on to the state of the art. Very much appreciated.
> 
> I provided you some additional feedback on Discord esp. in regard of discovery. Have you had a chance looking at it? If you prefer Feedback only via the PR, I can provide here too.

I'm sorry I don't use discord that much, I've missed your reply. I'll take a look ASAP

**Review by Lord-Grey — CHANGES_REQUESTED** [review]

**Comment by Lord-Grey** [comment]
@sanzoghenzo ssdp Discovery works now!

File: addon.xml
Comment (Lord-Grey) [inline addon.xml:8]: I am not sure, should it not be 

`<extension point="xbmc.service" library="addon.py"/>`

I guess there is no service.py? In the other addon this was changed that way...

**Review by Lord-Grey — CHANGES_REQUESTED** [review]

File: resources/settings.xml
Comment (sanzoghenzo) [reply resources/settings.xml:13]: I was used before to store the SSDP discovered USN, bus never read to do something meaningful. I'll remove it.

File: addon.xml
Comment (sanzoghenzo) [reply addon.xml:8]: That is the main entry point, and can be named as we want. The important thing is that the script in the root folder is named in the same way.

**Comment by sanzoghenzo** [comment]
Found some time to apply the changes!
I think we're all set now

**Review by Lord-Grey — APPROVED** [review]

**Comment by Lord-Grey** [comment]
Thanks for turning this round quickly!

I was wondering, if it makes sense already to include the multi-instance handling fix referred to in issue #18? I think the fix need to be amended a bit that as this scenario applies only to instance related components (but that I can provide).
Or should we better do an additional PR after the merge? What do you think?

**Comment by sanzoghenzo** [comment]
Sure we can add it!
What are the components that needs to be handled that way?

**Comment by Lord-Grey** [comment]
> What are the components that needs to be handled that way?

From the list of Components 
https://github.com/hyperion-project/hyperion.control/blob/87bb83f190d5fd048c29cfe0f6fb90f9b4b3d1b9/lib/utils.py#L40

number 0,1 and 5 are to be excluded from executing the command via all instances.
Those components are linked to the first instance only.

**Comment by sanzoghenzo** [comment]
Done.
 
> Edit: I just realized that the list even contains components that were deprecated and not include newer ones. If you included the instance handling, if will check what needs to be changed in the lists in the meantime.

I did got rid of deprecated entries earlier, but didn't add the new ones.

**Comment by Lord-Grey** [comment]
> I did get rid of deprecated entries earlier, but didn't add the new ones.

Great! I did not check your list :)
The additional Component is ["AUDIO" ](https://github.com/hyperion-project/hyperion.ng/blob/c0dc08b0c07e22f76ea2c88f5e3e18078ebcdc97/include/utils/Components.h#L84) which also is linked to Instance 0.
It is "Audio Capture" for the translation file.

**Comment by Lord-Grey** [comment]
Cool, you already envisioned the proper ordering I had in mind, too.
I will do some testing later…

**Comment by Lord-Grey** [comment]
@sanzoghenzo I am really sorry, it looks like only the forward is applicable to instance 0.

If you would not mind, please change https://github.com/sanzoghenzo/hyperion.control/blob/e66f09c2ff08a9254aff5d6ceec5b5d987d779f7/resources/lib/api_client.py#L79

to

` if component in {"FORWARDER"}:`

I mixed that the grabbers are only running ones, but the signal is piped per instance (component)...

Everything else is working great at my end!!!
You have done a tremendous job working refactoring the legacy code diligently.

Is everything ready to merge (after the small change above) from your end?

**Comment by sanzoghenzo** [comment]
> Is everything ready to merge (after the small change above) from your end?

Yes it is!

The token request thing can wait, it is better suited for a separate PR.

**Comment by Lord-Grey** [comment]
@sanzoghenzo I published the release, but the github actions jobs fails
https://github.com/hyperion-project/hyperion.control/actions/runs/5717816248/job/15492341002

Would you mind having a look, please?

**Pull request merged**
